src/sagemaker_xgboost_container/data_utils.py

```diff
@@ -505,33 +517,45 @@ def get_dmatrix(data_path, content_type, csv_weights=0, is_pipe=False):
     :param csv_weights: Only used if file_type is 'csv'.
                         1 if the instance weights are in the second column of csv file; otherwise, 0
     :param is_pipe: Boolean to indicate if data is being read in pipe mode
+    :param subsample_on_read: None or a value in (0, 1] to indicate how much of the dataset should
+            be read into memory.
     :return: xgb.DMatrix or None
     """
     if not (os.path.exists(data_path) or (is_pipe and os.path.exists(data_path + '_0'))):
         return None
-    else:
-        if os.path.isfile(data_path) or is_pipe:
-            files_path = data_path
-        elif not is_pipe:
-            for root, dirs, files in os.walk(data_path):
-                if dirs == []:
-                    files_path = root
-                    break
-        if content_type.lower() == CSV:
-            dmatrix = get_csv_dmatrix(files_path, csv_weights, is_pipe)
-        elif content_type.lower() == LIBSVM:
-            dmatrix = get_libsvm_dmatrix(files_path, is_pipe)
-        elif content_type.lower() == PARQUET:
-            dmatrix = get_parquet_dmatrix(files_path, is_pipe)
-        elif content_type.lower() == RECORDIO_PROTOBUF:
-            dmatrix = get_recordio_protobuf_dmatrix(files_path, is_pipe)
-
-        if dmatrix and dmatrix.get_label().size == 0:
-            raise exc.UserError(
-                "Got input data without labels. Please check the input data set. "
-                "If training job is running on multiple instances, please switch "
-                "to using single instance if number of records in the data set "
-                "is less than number of workers (16 * number of instance) in the cluster.")
+
+    if logging.isEnabledFor(logging.WARNING):
+        if subsample_on_read and (content_type.lower() != CSV or not is_pipe):
+            logging.warning(
+                "The subsample_on_read hyperparameter is only effective when "
+                "used in pipe mode with a CSV dataset.")
+
+
+    if os.path.isfile(data_path) or is_pipe:
+        files_path = data_path
+    elif not is_pipe:
+        for root, dirs, files in os.walk(data_path):
+            if dirs == []:
+                files_path = root
+                break
+    if content_type.lower() == CSV:
+        dmatrix = get_csv_dmatrix(files_path,
```

Comment:
There is value in having this subsampling for all formats. Is it possible to abstract the code to ensure it applies for all reads?

Reply:
I moved the subsampling logic into ML-IO. It means the logic is now available for all code paths that use ML-IO for data reading (as of today CSV/pipe and RecordIO-protobuf/[file, pipe])

Reply:
How does the subsampling interact with the batch size? I'm assuming the size of the example will still be equal to the batch size, but the reader will throw away however many records it reads based on the subsampling ratio until it reaches the batch size.